### PR TITLE
Update static graph docs

### PR DESCRIPTION
**Pull request by cdmihai**
The changes reflects the state of things with #5222's changes. As potential extra changes go into #5222, I'll also update this PR.

More specifically, #5222 implements this region of the docs:  https://github.com/microsoft/msbuild/compare/master...cdmihai:cacheAndIsolationDocumentation?expand=1#diff-5de82af4d8f52787d69b38fd58c509caR44

### CODE REVIEW

File: documentation/specs/static-graph.md [hunk 1]
```diff
@@ -297,56 +311,50 @@ Because referenced projects and their entry targets are guaranteed to be in the
...
 
-Input cache files constraints:
-- A ConfigCache / ResultsCache mapping must be unique between all input caches (multiple input caches cannot provide information for the same cache entry)
-- For each input cache, `ConfigCache.Entries.Size == ResultsCache.Entries.Size`
-- For each input cache, there is exactly one mapping from ConfigCache to ResultsCache
+A reference is exempt only in projects that add the reference in `GraphIsolationExemptReference`. If multiple projects need to exempt the same reference, all of them need to add the reference to `GraphIsolationExemptReference`.
 
-Output cache file constraints:
-- the output cache file contains results only for additional work performed in the current BeginBuild / EndBuild session. Entries from input caches are not transferred to the output cache.
+For now, self builds (a project building itself with different global properties) are also exempt from isolation constraints, but this behaviour is of dubious value and might be changed in the future.
```
Comment (Forgind) [inline documentation/specs/static-graph.md:355]:
```suggestion
For now, self-builds (a project building itself with different global properties) are also exempt from isolation constraints, but this behaviour is of dubious value and might be changed in the future.
```

File: documentation/specs/static-graph-implementation-details.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+# Single project isolated builds: implementation details
+
+<!-- workflow -->
+Single project isolated builds can be achieved by providing MSBuild with input and output cache files.
+
+The input cache files should contain the cached results all the targets that a project calls on its references. When a project executes, it will naturally build its references via [MSBuild task](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-task) calls. In isolated builds, the engine, instead of executing these tasks, will serve them from the provided input caches. In an isolated project build, only the current project should be building targets. Any other referenced projects should be provided form the input caches. 
+
+The output cache file tells MSBuild where it should serialize the results of the current project. This output cache would become an input cache for all other projects that depend on the current project.
+The output cache file can be omitted in which case the build would just reuse prior results but not write out any new results. This could be useful when one wants to replay a build from previous caches.
```
[the remark this answers is not in the document]
Comment (cdmihai) [reply documentation/specs/static-graph-implementation-details.md:14]: >By "replay a build from previous caches," do you mean going through the steps of a build as in seeing at what point it went wrong?

More like executing a project without executing its references. For example, the VS feature of re-building just one project could be implemented this way with caches.

[the remark this answers is not in the document]
Comment (cdmihai) [reply documentation/specs/static-graph-implementation-details.md:14]:
```suggestion
The output cache file can be omitted in which case the build reuses prior results but does not write out any new results. This is useful when one wants to re-execute the build for a project without building its references.
```

File: documentation/specs/static-graph-implementation-details.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+## Input / Output cache implementation
+<!-- cache structure -->
+The cache files contain the serialized state of MSBuild's [ConfigCache](https://github.com/Microsoft/msbuild/blob/master/src/Build/BackEnd/Components/Caching/ConfigCache.cs) and [ResultsCache](https://github.com/Microsoft/msbuild/blob/master/src/Build/BackEnd/Components/Caching/ResultsCache.cs). These two caches have been traditionally used by the engine to cache build results. For example, it is these caches which ensure that a target is only built once per build submission. The `ConfigCache` entries are instances of [BuildRequestConfiguration](https://github.com/microsoft/msbuild/blob/37c5a9fec416b403212a63f95f15b03dbd5e8b5d/src/Build/BackEnd/Shared/BuildRequestConfiguration.cs#L25). The `ResultsCache` entries are instances of [BuildResult](https://github.com/microsoft/msbuild/blob/37c5a9fec416b403212a63f95f15b03dbd5e8b5d/src/Build/BackEnd/Shared/BuildResult.cs#L34), which compose instances of [TargetResult](https://github.com/microsoft/msbuild/blob/37c5a9fec416b403212a63f95f15b03dbd5e8b5d/src/Build/BackEnd/Shared/TargetResult.cs#L22).
```
[the remark this answers is not in the document]
Comment (cdmihai) [reply documentation/specs/static-graph-implementation-details.md:20]:
```suggestion
The cache files contain the serialized state of MSBuild's [ConfigCache](https://github.com/Microsoft/msbuild/blob/master/src/Build/BackEnd/Components/Caching/ConfigCache.cs) and [ResultsCache](https://github.com/Microsoft/msbuild/blob/master/src/Build/BackEnd/Components/Caching/ResultsCache.cs). These two caches have been traditionally used by the engine to cache build results. For example, it is these caches which ensure that a target is only built once per build submission. The `ConfigCache` entries are instances of [BuildRequestConfiguration](https://github.com/microsoft/msbuild/blob/37c5a9fec416b403212a63f95f15b03dbd5e8b5d/src/Build/BackEnd/Shared/BuildRequestConfiguration.cs#L25). The `ResultsCache` entries are instances of [BuildResult](https://github.com/microsoft/msbuild/blob/37c5a9fec416b403212a63f95f15b03dbd5e8b5d/src/Build/BackEnd/Shared/BuildResult.cs#L34), which contain or more instances of [TargetResult](https://github.com/microsoft/msbuild/blob/37c5a9fec416b403212a63f95f15b03dbd5e8b5d/src/Build/BackEnd/Shared/TargetResult.cs#L22). 
```

File: documentation/specs/static-graph.md [hunk 4]
```diff
@@ -297,56 +311,50 @@ Because referenced projects and their entry targets are guaranteed to be in the
...
 
-Furthermore, running in this mode enforces that each (project, global properties) pair is executed only once and must execute all targets needed by all projects which reference that node. This gives it a concrete start and end time, which leads to some perf optimizations, like garbage collecting all project state (except the build results) once it finishes building. This can greatly reduce the memory overhead for large builds.
+Furthermore, running in this mode enforces that each (project, global properties) pair is executed only once and must execute all targets needed by all projects which reference that node. This gives it a concrete start and end time, which leads to some potential perf optimizations, like garbage collecting all project state (except the build results) once it finishes building. This can greatly reduce the memory overhead for large builds.
```
[the remark this answers is not in the document]
Comment (cdmihai) [reply documentation/specs/static-graph.md:314]: I'll do a pass and try to reduce duplicates.

[the remark this answers is not in the document]
Comment (cdmihai) [reply documentation/specs/static-graph.md:314]: Upon looking over it, not much can be reduced. My intention was to keep high level information in this document, and move all code implementation details into the other one.

[the remark this answers is not in the document]
Comment (Forgind) [reply documentation/specs/static-graph.md:314]: Ok, sounds good. My best idea for how to completely eliminate any duplicate information would also be a nightmare to maintain, so... 😄

File: documentation/specs/static-graph-implementation-details.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Forgind) [reply documentation/specs/static-graph-implementation-details.md]: I like that.

**Pull request closed without merging**